extended/src/main/java/io/kubernetes/client/extended/controller/DefaultController.java

```diff
@@ -0,0 +1,235 @@
+      log.error("Controller {} failed pre-run check, exiting..", this.name);
+      return;
+    }
+
+    // spawns worker threads for the controller.
+    CountDownLatch latch = new CountDownLatch(workerCount);
+    for (int i = 0; i < this.workerCount; i++) {
+      final int workerIndex = i;
+      workerThreadPool.scheduleWithFixedDelay(
+          () -> {
+            if (log.isDebugEnabled()) {
+              log.debug("Starting controller {} worker {}..", this.name, workerIndex);
+            }
+            this.worker();
+            latch.countDown();
+            if (log.isDebugEnabled()) {
+              log.debug("Exiting controller {} worker {}..", this.name, workerIndex);
+            }
+          },
+          0,
+          1,
+          TimeUnit.SECONDS);
+    }
+    try {
+      if (log.isDebugEnabled()) {
+        log.debug("Controller {} bootstrapping..", this.name);
+      }
+      latch.await();
+    } catch (Throwable t) {
+      log.error("Controller {} hitting exception.", this.name);
+      log.error("Aborting controller.", t);
+    } finally {
+      log.info("Controller {} exited", this.name);
+    }
+  }
+
+  @Override
+  public void shutdown() {
+    // shutdown work-queue before the thread-pool
+    workQueue.shutDown();
+    workerThreadPool.shutdown();
+  }
+
+  private void worker() {
+    // taking tasks from work-queue in a loop
+    while (!workQueue.isShuttingDown()) {
+      Reconciler.Request request = null;
+      try {
+        request = workQueue.get();
+      } catch (InterruptedException e) {
+        // we're reaching here mostly because of forcibly shutting down the controller.
+        log.error("Controller worker interrupted.. keeps working until work-queue shutdown", e);
+        if (request != null) {
+          workQueue.addRateLimited(request);
+        }
+      }
+      // request is expected to be null, when the work-queue is shutting-down.
+      if (request == null) {
+        log.info("Controller {} worker exiting because work-queue has shutdown..");
+        return;
+      }
+      if (log.isDebugEnabled()) {
+        log.debug("Controller {} start reconciling {}..", request);
+      }
+
+      // do reconciliation, invoke user customized logic.
+      Reconciler.Result result = this.reconciler.reconcile(request);
+
+      // checks whether do a re-queue (on failure)
+      if (result.isRequeue()) {
+        if (result.getRequeueAfter() == null) {
+          if (log.isDebugEnabled()) {
+            log.debug("Controller {} reconciling {} failed, requeuing {}..", request);
+          }
+          workQueue.addRateLimited(request);
+        } else {
+          if (log.isDebugEnabled()) {
+            log.debug(
+                "Controller {} reconciling {} failed, requeuing after {}..",
+                request,
+                result.getRequeueAfter());
+          }
+          workQueue.addAfter(request, result.getRequeueAfter());
+        }
+      }
+
+      if (request != null) {
```

Comment:
How could request be null here? There seems to be lots of usage of request above that would fail if request is null...

Reply:
There appears to be a null check on line 150....

Reply:
uh indeed.. the null check is redandunt here